models/buffer.go

```diff
@@ -12,18 +10,36 @@ var (
 	AgentMetricsDropped = selfstat.Register("agent", "metrics_dropped", map[string]string{})
 )
 
-// Buffer stores metrics in a circular buffer.
-type Buffer struct {
-	sync.Mutex
-	buf   []telegraf.Metric
-	first int // index of the first/oldest metric
-	last  int // one after the index of the last/newest metric
-	size  int // number of metrics currently in the buffer
-	cap   int // the capacity of the buffer
+const (
+	StrategyMemory   = "memory"
+	StrategyDisk     = "disk"
+	StrategyOverflow = "overflow"
+)
+
+type Buffer interface {
+
+	// Len returns the number of metrics currently in the buffer.
+	Len() int
+
+	// Add adds metrics to the buffer and returns number of dropped metrics.
+	Add(metrics ...telegraf.Metric) int
 
-	batchFirst int // index of the first metric in the batch
-	batchSize  int // number of metrics currently in the batch
+	// Batch returns a slice containing up to batchSize of the oldest metrics not
+	// yet dropped.  Metrics are ordered from oldest to newest in the batch.  The
+	// batch must not be modified by the client.
+	Batch(batchSize int) []telegraf.Metric
 
+	// Accept marks the batch, acquired from Batch(), as successfully written.
+	Accept(metrics []telegraf.Metric)
+
+	// Reject returns the batch, acquired from Batch(), to the buffer and marks it
+	// as unsent.
+	Reject([]telegraf.Metric)
+}
+
+// BufferMetrics holds common metrics used for buffer implementations.
+// Implementations of Buffer should embed this struct in them.
+type BufferMetrics struct {
```

Comment:
Maybe
```suggestion
type BufferStats struct {
```
as those are only stats now...